NDTensors/src/tensor/tensor.jl

```diff
@@ -184,6 +188,8 @@ setstorage(T, nstore) = tensor(nstore, inds(T))
 
 setinds(T, ninds) = tensor(storage(T), ninds)
 
+iszerodata(t::Tensor) = data(t) isa NDTensors.Zeros
```

Comment:
I think this would be better:
```julia
iszerodata(t::Tensor) = iszerodata(storage(t))
iszerodata(s::TensorStorage) = iszerodata(data(s))
iszerodata(::NDTensors.Zeros) = true
iszerodata(::Any) = false
```

Also I think `iszerodata` doesn't sound quite right for the name of this function, since a `Tensor` might have Zeros data, but it isn't itself Zeros data (or data at all, by our terminology). Also `zero` seems too general since there could be tensors with allocated numerical zeros. Maybe `is_unallocated_zero`?

We could design this around a more generic function:
```julia
data_isa(t::Tensor, datatype::Type) = data_isa(storage(t), datatype)
data_isa(s::TensorStorage, datatype::Type) = data_isa(data(s), datatype)
data_isa(d::AbstractArray, datatype::Type) = d isa datatype # Might have to unwrap if it is reshaped, sliced, etc.
```
and then define:
```julia
is_unallocated_zero(a) = data_isa(a, NDTensors.Zeros)
```
This could generalize to a function `storage_isa` which could be used to define `iscombiner`, `isdense`, etc.

Reply:
I really like the `data_isa` function, its again a home brewed version trait function but I think its good when we can so dynamically define what a datatype is in the backend. I have implemented them in the code and made some updates